### PR TITLE
[WIP] NER Enhancement

### PR DESCRIPTION
**Pull request by SSaishruthi**
Added code to,

- store checkpoints
- load checkpoints
- batch inference 
- f1 score

Note and questions:

- Using the same MAX parameters.
- 300d vector
- Added F1 score from addons. Should we have this? If yes, I might need to update this to the latest version as I have created F1 PR to addons the last year-end. 
- Should I add sentence sorting for batch inference?

### CODE REVIEW

**Comment by SSaishruthi** [comment]
cc @frreiss

**Comment by MLnick** [comment]
@SSaishruthi I believe the tests need to be updated to reflect the new batch input (list of string)?

**Comment by SSaishruthi** [comment]
@MLnick Updated test and result fields

File: core/model.py [hunk 1]
```diff
@@ -58,26 +57,47 @@ def _load_assets(self, path):
         self.proc_fn = get_processing_word(vocab_words, vocab_chars, lowercase=True, chars=True)
 
         self.id_to_tag = {idx: v for v, idx in vocab_tags.items()}
+        self.dict_vocal_tags_len = len(vocab_tags)
+        self.id_to_tag.update({self.dict_vocal_tags_len: 'O'})
```
Comment (MLnick) [inline core/model.py:61]: Is this really required?

I guess its for the `_post_process` to only return the predicted tags for input words, not for padding?

You should be able to use `pad_tag` to filter?

File: training/training_code/f1.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
Comment (MLnick) [inline training/training_code/f1.py:1]: As a general comment if this is the same as what is in `tf.addons`, we should prefer to use the one from `tf.addons`

File: training/training_code/f1.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+                dtype=self.dtype)
+
+        self.true_positives = _zero_wt_init('true_positives')
+        self.false_positives = _zero_wt_init('false_positives')
+        self.false_negatives = _zero_wt_init('false_negatives')
+        self.weights_intermediate = _zero_wt_init('weights_intermediate')
+
+    # TODO: Add sample_weight support, currently it is
+    # ignored during calculations.
+    def update_state(self, y_true, y_pred, sample_weight=None):
```
Comment (MLnick) [inline training/training_code/f1.py:111]: I am pretty sure that supporting the `sample_weight` mask is required for our case in order to get a proper result - otherwise it's computing the score including padding...

File: training/training_code/train_ner.py [hunk 4]
```diff
@@ -166,14 +169,38 @@ def _build_model(embedding_weights, char_bidirectional=False, concat_bidirection
...
+
+
+callbacks = ModelCheckpoint(filepath=checkpoint_path,
+                            period=1,
+                            verbose=1,
+                            save_weights_only=True,
+                            )
+
+'''
+# Load latest checkpoints and restart training
```
Comment (MLnick) [inline training/training_code/train_ner.py:195]: Why is this commented out?

File: training/training_code/train_ner.py [hunk 5]
```diff
@@ -23,9 +23,11 @@
 from tensorflow.python.keras.layers import concatenate
 from tensorflow.python.keras.models import Model
 from tensorflow.python.keras.optimizers import Adam
-
+from tensorflow.python.keras.callbacks import ModelCheckpoint
```
Comment (MLnick) [inline training/training_code/train_ner.py:26]: Have we checked that the new model training code still works on WML - since we still need to target it.

Specifically, though this should be TF 2.x compatible code, does it still work on 1.15 for WML?

File: core/model.py (hunk 1 above)
Comment (SSaishruthi) [reply core/model.py:61]: I was getting an error when using the V1  dataset as prediction number had `pad_tag`. I have code to update `id_to_tag` as well as remove tag during the post-processing step.

File: training/training_code/train_ner.py (hunk 4 above)
Comment (SSaishruthi) [reply training/training_code/train_ner.py:195]: I have commented this thinking we can use this part of the code only if checkpoints are available / want to start from the latest checkpoints.

File: training/training_code/f1.py (hunk 3 above)
Comment (SSaishruthi) [reply training/training_code/f1.py:111]: we don't have `sample_weight` in addons. I wrote the first version of this metric and added it to addons. I did not include sample weights. I don't think we anyone took that up and updated the code.

File: training/training_code/train_ner.py (hunk 5 above)
Comment (SSaishruthi) [reply training/training_code/train_ner.py]: I tested locally with TF 2.1 and TF 1.15.

I did not check with WML. Will run the WML test as well
